**Generalized Additive Models** (GAMs) are a flexible family of models that do not assume a linear relationship between the response and predictor variables (39). The impact of any predictor variable is captured through a so-called smooth function which, depending on the underlying patterns in the data, can be either linear or nonlinear (1). To balance the overfitting/underfitting trade-off, GAMs allow both the smoothness (i.e., how wiggly a given term is) and the effect size of component functions to be controlled through regularization/penalization (1, 2).

We constructed three hierarchical models to describe temporal trends at the **(i)** population level (model P); **(ii)** population and the social group level (model P+G); and **(iii)** population, the social group level, and the individual host level (model P+G+H). The modeled data are hierarchical as microbiome samples can be grouped into those collected from individual hosts, which themselves can be further grouped into social groups, with variables measured for both hosts, social groups and the population as a whole. GAMs were fitted to a subset of the full data shown in fig. S2, consisting of 4,277 16S rRNA gene sequencing profiles from the best-sampled 56 baboons living in the 5 main social groups between 2002 and 2010 (min=48; median = 72.5; max = 164 samples). We fit GAMs using the R function bam() from the package mgcv which uses fast restricted maximum likelihood (method=”fREML”) as the smoothing parameter estimation method (1). We fitted each GAM (i.e., model P; model P+G; and model P+G+H) to 52 different gut microbiome features describing different aspects of the gut microbiome: three principal components (PC1-PC3) to describe community-level changes, three alpha diversity indices (species richness, exponent of Shannon diversity, and the inverse Simpson index), and clr-transformed relative abundances of 12 phyla and 34 families. The results of these models are presented in fig. 4A and tables S5.

To minimize residual autocorrelation, we used the residuals of a Gaussian process as the response variable in each GAM. More specifically, this was done by first making each of the modeled host’s time series (fig. S2) continuous by ‘filling in’ missing collection dates (i.e. days) with NAs, and then fitting a Gaussian process via the R function gausspr() from the package kernlab (3). The input to the gausspr() function was a given host’s time series containing the any microbial feature. We then extracted and used the residuals from the Gaussian process as the response variable in each GAM.

References
1.  S. N. Wood, Generalized Additive Models: An Introduction with R, Second Edition (CRC Press, 2017).
2.	E. J. Pedersen, D. L. Miller, G. L. Simpson, N. Ross, Hierarchical generalized additive models in ecology: an introduction with mgcv. PeerJ 7, e6876 (2019).
3.	A. Karatzoglou, A. Smola, K. Hornik, A. Zeileis, kernlab- AnS4Package for Kernel Methods inR. J. Stat. Softw. 1, 1–20 (2004).
